src/core/Akka/IO/TcpConnection.cs

```diff
@@ -778,110 +818,191 @@ public WriteFileFailed(SocketException cause)
+            /// Thrown when data to buffer is larger then allowed <see cref="_maxQueueSizeInBytes"/>
+            /// </exception>
+            public void EnqueueSimpleWrites(WriteCommand command, IActorRef sender)
+            {
+                foreach (var writeInfo in ExtractFromCommand(command))
                 {
-                    return _connection.Socket.Send(data.Buffers);
+                    var sizeAfterAppending = _totalSizeInBytes + writeInfo.DataSize;
+                    if (_maxQueueSizeInBytes.HasValue && _maxQueueSizeInBytes.Value < sizeAfterAppending)
+                    {
+                        throw new InternalBufferOverflowException($"Could not receive write command of size {writeInfo.DataSize} bytes, " +
+                                                                  $"because buffer limit is {_maxQueueSizeInBytes} bytes and " +
+                                                                  $"it is already {_totalSizeInBytes} bytes");
+                    }
+
+                    _totalSizeInBytes = sizeAfterAppending;
+                    _queue.Enqueue((writeInfo.Command, sender, writeInfo.DataSize));
                 }
-                catch (SocketException e) when (e.SocketErrorCode == SocketError.WouldBlock)
+            }
+            
+            /// <summary>
+            /// Adds all <see cref="SimpleWriteCommand"/> subcommands stored in provided command.
+            /// Performs buffer size checks for all, except first one, that is not buffered
+            /// </summary>
+            /// <returns>
+            /// Not buffered (and not checked) first <see cref="SimpleWriteCommand"/>
+            /// </returns>
+            /// <exception cref="InternalBufferOverflowException">
+            /// Thrown when data to buffer is larger then allowed <see cref="_maxQueueSizeInBytes"/>
+            /// </exception>
+            public SimpleWriteCommand EnqueueSimpleWritesExceptFirst(WriteCommand command, IActorRef sender)
+            {
+                SimpleWriteCommand first = null;
+                foreach (var writeInfo in ExtractFromCommand(command))
                 {
-                    try
+                    if (first == null)
                     {
-                        _connection.Socket.Blocking = true;
-                        return _connection.Socket.Send(data.Buffers);
+                        first = writeInfo.Command;
+                        continue;
                     }
-                    finally
+                    
+                    var sizeAfterAppending = _totalSizeInBytes + writeInfo.DataSize;
+                    if (_maxQueueSizeInBytes.HasValue && _maxQueueSizeInBytes.Value < sizeAfterAppending)
                     {
-                        _connection.Socket.Blocking = false;
+                        throw new InternalBufferOverflowException($"Could not receive write command of size {writeInfo.DataSize} bytes, " +
```

Comment:
I wouldn't throw here - I'd just skip any additional writes once the buffer is full enough. Pick it up on the next pass.

The only time we should throw is if a write is so large that it could never possibly fit in the socket buffer in the first place - does that make sense?

Reply:
Or am I misunderstanding the code here?

Reply:
Mm... The idea is that if (this is optional) `TcpConnection`'s internal writes queue is limited in size, we should start dropping incoming writes. 
As akka's docs says, there are few strategies do deal with such overflows. Basically, your actor either is waiting for ack messages before send next write, either use auto-write-suspending thing and just write until getting `CommandFailed` response, after which wait for `WritingResumed` message and go on.

So, here I was throwing exception to catch it with `TcpConnection` actor and send `CommandFailed` (+ suspend writing). Now it is just returning `bool` instead of throwing, you are absolutely write, no reason to throw here.

Also, if incoming write is too large, but we can send it immediately, we will not give an error because there is no particular limit for buffers in `SocketAsyncEventArgs`. But yes, if we have to buffer write and it is too big and hits the limit - we will... not throw now, but just get `false` response from internal `PendingSimpleWritesQueue` and report failure to the sender.

Did I got your point here?

Reply:
> Also, if incoming write is too large, but we can send it immediately, we will not give an error because there is no particular limit for buffers in SocketAsyncEventArgs. But yes, if we have to buffer write and it is too big and hits the limit - we will... not throw now, but just get false response from internal PendingSimpleWritesQueue and report failure to the sender.

Yeah, that sounds right - we're never going to be able to fit a 4GB message into a 256k buffer, so we always need to reject that. But otherwise we can buffer and wait. Sounds good.